azkaban-web-server/src/main/java/azkaban/webapp/servlet/ProjectManagerServlet.java

```diff
@@ -1938,18 +1938,20 @@ private void ajaxHandleUpload(final HttpServletRequest req, final HttpServletRes
       out = new BufferedOutputStream(new FileOutputStream(archiveFile));
       IOUtils.copy(item.getInputStream(), out);
       out.close();
-      if (this.enableQuartz) {
-        //todo chengren311: should maintain atomicity,
-        // e.g, if uploadProject fails, associated schedule shouldn't be added.
-        this.scheduler.unschedule(project);
-      }
 
       // get the locked flows for the project, so that they can be locked again after upload
       final List<Pair<String, String>> lockedFlows = getLockedFlows(project);
+      // record the existing project flows before upload
+      final List<Flow> existingFlows = project.getFlows();
 
+      // validate project zip's dependencies and persist the new project metadata into DB
       final Map<String, ValidationReport> reports = this.projectManager
           .uploadProject(project, archiveFile, lowercaseExtension, user, props, uploaderIPAddr);
+
+      // Post-processing after upload
+      // reschedule data triggers if quartz is enabled
       if (this.enableQuartz) {
+        this.scheduler.unschedule(existingFlows, projectName);
```

Comment:
What will happen if project is uploaded successfully, then a flow trigger arrives before `unschedule` is called under these two cases?
1. Flow exists
2. Flow is renamed or deleted

Reply:
what do you mean by flow trigger arrives?
I assume you're asking when flow trigger is working at the moment of upload?
The flowTrigger system is using Quartz library, it does not have the concept of "flow". What we do is we define the name of a QuartzJob using project.flowName. 
So unschedule means stop that running QuartzJob and delete it forever.
1. if flow exists;
If we schedule again, the flowTrigger itself has the logic to guarantee when a flow trigger is added/scheduled, it immediately checks availability and adding this job to the routine kafka listener thread.  

2. if flow renamed/deleted
the old flow triggers is stopped and removed anyways.

Reply:
I believe Binyao is asking what if before `unschedule` is called, the QuartzJob is triggered.

Reply:
> I believe Binyao is asking what if before unschedule is called, the QuartzJob is triggered.

Yes. that's my Q.

Reply:
each QuartzJob are scheduled by Quartz library based on cron expression, but the details like ExecutionContext is controlled by us (refer to FlowTriggerService Class), we store all the information in JobDetails. FlowTriggerService would start/cancel trigger, which has our self-defined triggerInstance which contains lots of dali dependencies, each triggerInstance we store Project Object directly within it, and when triggerExecution happens, we use the stored project data to start execution, so if new project just updated in DB it will not change flow trigger's project data. 
 
If the flow has been renamed/deleted, but all dali dependencies are available so we're ready to executeFlow, the execution would try to start using old project/flows. But whether it's success to execute or not is not important and this should not affect uploading process. 
So even if flow was deleted/renamed, the schedule new trigger instance would succeed because JobKey contains flowName is different.

Reply:
but I think if we want to prevent the execution on old flow, we might need to change project cache a bit: in order to always reflecting new project metadata right after persist project in DB.
then change the fetch project in triggerInstance to use ProjectId to get most updated record.
This might be a big change I wonder if it's worth the effort.

Reply:
That's nice to hear. Thanks! I think executing old flows is fine (good workaround already).